File: client/files2/Área de Trabalho/LP-codes/test/c-python/binarie-trees/b-trees.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

class BinaryTree:
    def __init__(self):
        self.root = None
    
    def insert(self, value):
        if not self.root:
            self.root = Node(value)
        else:
            self._insert_recursive(self.root, value)
    
    def _insert_recursive(self, current_node, value):
        if value < current_node.value:
            if current_node.left is None:
                current_node.left = Node(value)
            else:
                self._insert_recursive(current_node.left, value)
        else:
            if current_node.right is None:
                current_node.right = Node(value)
            else:
                self._insert_recursive(current_node.right, value)
    
    def search(self, value):
        return self._search_recursive(self.root, value)
    
    def _search_recursive(self, current_node, value):
        if current_node is None:
            return False
        if current_node.value == value:
            return True
        elif value < current_node.value:
            return self._search_recursive(current_node.left, value)
        else:
            return self._search_recursive(current_node.right, value)

    def inorder_traversal(self):
        result = []
        self._inorder_recursive(self.root, result)
        return result
    
    def _inorder_recursive(self, current_node, result):
        if current_node:
            self._inorder_recursive(current_node.left, result)
            result.append(current_node.value)
            self._inorder_recursive(current_node.right, result)

# Exemplo de uso
tree = BinaryTree()
tree.insert(5)
tree.insert(3)
tree.insert(7)
tree.insert(2)
tree.insert(4)

print(tree.search(4))  # True
print(tree.search(6))  # False

print(tree.inorder_traversal())  # [2, 3, 4, 5, 7]

File: client/files2/Área de Trabalho/LP-codes/test/c-python/binarie-trees/b-trees.rs
use std::env;

// Definição da estrutura do nó da árvore binária
struct TreeNode {
    value: i32,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

// Função para criar a árvore binária com a profundidade dada
fn create_binary_tree(depth: i32, value: i32) -> Option<Box<TreeNode>> {
    if depth == 0 {
        None
    } else {
        Some(Box::new(TreeNode {
            value,
            left: create_binary_tree(depth - 1, value * 2),
            right: create_binary_tree(depth - 1, value * 2 + 1),
        }))
    }
}

// Função para realizar a travessia em ordem (inorder traversal) na árvore binária
fn inorder_traversal(node: &Option<Box<TreeNode>>) {
    if let Some(inner) = node {
        inorder_traversal(&inner.left);
        println!("{}", inner.value);
        inorder_traversal(&inner.right);
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();

    if args.len() != 2 {
        println!("Uso: {} <profundidade>", args[0]);
        return;
    }

    let depth: i32 = args[1].parse().expect("Profundidade inválida");
    let root = create_binary_tree(depth, 1);

    println!("Árvore binária criada com profundidade {}", depth);
    println!("Resultado da travessia em ordem (inorder traversal):");
    inorder_traversal(&root);
}
